src/utils/commands/command_util.py

```diff
@@ -109,82 +109,66 @@ def rank(self, discord_id: str) -> str:
+        database for completions and updates via leetcode API ONLY if necessary. Gathers the list
         of current weekly challenges from the database.
         """
         result = ""
         user = self.database.users.load_by_discord_id(discord_id)
         if user:
-            leetcode_user_id = user["leetcode_id"]
             challenge = self.database.weekly_challenges.get_latest()
-            if not challenge:
-                result += "No current challenges"
-            else:
-                questions = self.database.weekly_questions.load_by_challenge_id(challenge["id"])
-                total = len(questions)
-                if total == 0:
-                    result += "No current weekly questions"
-                else:
-                    completed = 0
-                    for question in questions:
-                        question["complete"] = self.leetcode.check_challenge_completion(
-                            leetcode_user_id, question["title_slug"]
-                        )
-                        if question["complete"]:
-                            completed += 1
-                    percentage = int(completed/total * 100)
-                    result += f"User {leetcode_user_id}'s Weekly Challenge status: {percentage}%\n"
-                    for question in questions:
-                        complete = question["complete"]
-                        result += (
-                            f"\t{self.reaction_complete if complete else self.reaction_incomplete}"
-                            f"\t-\t{question['title']}\n"
+            if challenge:
+                user_leetcode_id = user["leetcode_id"]
+                self.update_user_completions(user_leetcode_id, challenge["id"])
+                completions = self.database.question_completions.load_all_title_slugs_by_user(
+                    user_leetcode_id
+                    )
+                question_slugs = self.database.weekly_questions.load_all_title_slugs_by_challenge(
+                    challenge["id"]
+                    )
+                completed_in_database = list(set(completions) & set(question_slugs))
+                total_completions = len(completed_in_database)
+                total_questions = len(question_slugs)
+                percentage = int(total_completions / total_questions * 100)
+                result += f"User {user_leetcode_id}'s Weekly Challenge status: {percentage}%\n"
+                completed_all = total_completions == total_questions
+                for question_slug in question_slugs:
+                    complete = True if completed_all else \
+                    self.database.question_completions.check_completion(
+                        user_leetcode_id,
+                        question_slug
                         )
+                    question = self.database.weekly_questions.load_by_title_slug(question_slug)
```

Comment:
This seems like a redundant Database call. We could already have all of the weekly questions if we loaded them all earlier instead of loading by title slug. It would be better to make a new list of `question_slugs` derived from the original list of `questions`. Something like `question_slugs = [q['title_slug'] for q in questions]`.

Reply:
Great call, I've updated and pushed the change for this.